File: app/src/main/java/com/udacity/maluleque/meutako/ReportFragment.java
package com.udacity.maluleque.meutako;

import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;

import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.components.Description;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.formatter.PercentFormatter;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import com.github.mikephil.charting.utils.ColorTemplate;
import com.github.mikephil.charting.utils.MPPointF;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.udacity.maluleque.meutako.model.Transaction;
import com.udacity.maluleque.meutako.utils.DateUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import butterknife.BindView;
import butterknife.ButterKnife;


public class ReportFragment extends Fragment {
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";
    private static final String TAG = "ReportFragment";

    @BindView(R.id.pieChart)
    PieChart pieChart;
    @BindView(R.id.lineChart)
    LineChart lineChart;
    @BindView(R.id.barChart)
    BarChart barChart;

    private int mParam1;
    private String dataMonth;
    private FirebaseFirestore db;
    private ListenerRegistration registration;
    private FirebaseUser user;


    public ReportFragment() {

    }


    public static ReportFragment newInstance(int param1, String param2) {
        ReportFragment fragment = new ReportFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    public static <T> Predicate<T> distinctByKey(Function<? super T, ?> keyExtractor) {
        Set<Object> seen = ConcurrentHashMap.newKeySet();
        return t -> seen.add(keyExtractor.apply(t));
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getInt(ARG_PARAM1);
            dataMonth = getArguments().getString(ARG_PARAM2);
        }
        db = FirebaseFirestore.getInstance();
        user = FirebaseAuth.getInstance().getCurrentUser();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_report, container, false);
        ButterKnife.bind(this, view);

        getTransactions(dataMonth);
        return view;

    }

    void getTransactions(String dataMonth) {
        long[] dateIntervals = DateUtils.getDateIntervals(dataMonth);
        Query query = db.collection("users")
                .document(user.getUid())
                .collection("transactions")
                .whereGreaterThanOrEqualTo("date", dateIntervals[0])
                .whereLessThanOrEqualTo("date", dateIntervals[1])
                .orderBy("date", Query.Direction.DESCENDING);

        registration = query.addSnapshotListener(
                (queryDocumentSnapshots, e) -> {

                    if (e != null) {
                        Log.e(TAG, "Listen failed.", e);
                        return;
                    }

                    List<Transaction> transactions = new ArrayList<>();
                    for (QueryDocumentSnapshot doc : queryDocumentSnapshots) {
                        transactions.add(doc.toObject(Transaction.class));
                    }

                    if (transactions.isEmpty()) {
                        return;
                    }

                    showData(transactions);


                });

    }

    private void showData(List<Transaction> transactions) {
        setPieChartData(transactions);
        setBarData(transactions);
        setLine2ChartData(transactions);
    }

    private void setBarData(List<Transaction> transactions) {
        List<BarEntry> entries = new ArrayList<>();
        List<BarEntry> entriesExpense = new ArrayList<>();

        Map<String, Double> sumByExpenses = transactions.stream().filter(transaction -> transaction.getType().equals("Expense"))
                .collect(Collectors.groupingBy(Transaction::getType, Collectors.summingDouble(Transaction::getAmount)));

        Map<String, Double> sumByIncomes = transactions.stream().filter(transaction -> transaction.getType().equals("Income"))
                .collect(Collectors.groupingBy(Transaction::getType, Collectors.summingDouble(Transaction::getAmount)));

        for (Map.Entry<String, Double> transaction : sumByExpenses.entrySet()) {
            entries.add(new BarEntry(0, new Double(transaction.getValue()).floatValue()));
        }


        for (Map.Entry<String, Double> transaction : sumByIncomes.entrySet()) {
            entriesExpense.add(new BarEntry(2, new Double(transaction.getValue()).floatValue()));
        }


        BarDataSet set = new BarDataSet(entries, "Total Income");
        set.setColor(Color.GREEN);
        BarDataSet set2 = new BarDataSet(entriesExpense, "Total Expense");
        set2.setColor(Color.RED);

        BarData data = new BarData(set, set2);
        data.setBarWidth(0.9f); // set custom bar width
        Description description = new Description();
        description.setText("Totals expenses and incomes");
        barChart.setDescription(description);
        barChart.setData(data);
        barChart.setFitBars(true); // make the x-axis fit exactly all bars
        barChart.invalidate(); // refresh
    }

    private void setPieChartData(List<Transaction> transactions) {
        ArrayList<PieEntry> entries = new ArrayList<>();

        Map<String, Double> sum = transactions.stream().filter(transaction -> transaction.getType().equals("Expense")).collect(Collectors.groupingBy(Transaction::getCategory, Collectors.summingDouble(Transaction::getAmount)));
        sum.entrySet().forEach(x -> entries.add(new PieEntry(x.getValue().floatValue(), x.getKey())));

        PieDataSet dataSet = new PieDataSet(entries, "Categories");

        dataSet.setDrawIcons(false);
        dataSet.setSliceSpace(3f);
        dataSet.setIconsOffset(new MPPointF(0, 40));
        dataSet.setSelectionShift(5f);

        // add a lot of colors
        ArrayList<Integer> colors = new ArrayList<>();
        for (int c : ColorTemplate.VORDIPLOM_COLORS)
            colors.add(c);

        for (int c : ColorTemplate.JOYFUL_COLORS)
            colors.add(c);

        for (int c : ColorTemplate.COLORFUL_COLORS)
            colors.add(c);

        for (int c : ColorTemplate.LIBERTY_COLORS)
            colors.add(c);

        for (int c : ColorTemplate.PASTEL_COLORS)
            colors.add(c);

        colors.add(ColorTemplate.getHoloBlue());

        dataSet.setColors(colors);
        dataSet.setSelectionShift(0f);

        PieData data = new PieData(dataSet);
        data.setValueFormatter(new PercentFormatter(pieChart));
        data.setValueTextSize(11f);
        data.setValueTextColor(R.color.colorPrimaryText);
        pieChart.setData(data);

        // undo all highlights
        pieChart.highlightValues(null);
        pieChart.setCenterText("Expenses by category");
        Description description = new Description();
        description.setText("Check where you expend the most");
        pieChart.setDescription(description);
        pieChart.invalidate();
    }

    private void setLine2ChartData(List<Transaction> transactions) {

        Map<String, Double> sumByTypeAndDate = transactions.stream().filter(transaction -> transaction.getType().equals("Expense"))
                .collect(Collectors.groupingBy(Transaction::getFormattedDate, Collectors.summingDouble(Transaction::getAmount)));

        Map<String, Double> sumByIncomeAndDate = transactions.stream().filter(transaction -> transaction.getType().equals("Income"))
                .collect(Collectors.groupingBy(Transaction::getFormattedDate, Collectors.summingDouble(Transaction::getAmount)));


        ArrayList<Entry> values1 = new ArrayList<>();
        int ds = 0;

        for (Map.Entry<String, Double> transaction : sumByTypeAndDate.entrySet()) {
            values1.add(new Entry(ds, new Double(transaction.getValue()).floatValue()));
            ds = ds + 1;
        }


        ArrayList<Entry> values2 = new ArrayList<>();

        for (Map.Entry<String, Double> transaction : sumByIncomeAndDate.entrySet()) {
            values2.add(new Entry(ds, new Double(transaction.getValue()).floatValue()));
            ds = ds + 1;
        }

        LineDataSet setComp1 = new LineDataSet(values1, "Expenses");
        setComp1.setAxisDependency(YAxis.AxisDependency.LEFT);
        setComp1.setColor(Color.RED);
        setComp1.setCircleColor(Color.WHITE);

        LineDataSet setComp2 = new LineDataSet(values2, "Incomes");
        setComp2.setAxisDependency(YAxis.AxisDependency.LEFT);
        setComp2.setColor(Color.GREEN);
        setComp2.setCircleColor(Color.GREEN);

        List<ILineDataSet> dataSets = new ArrayList<ILineDataSet>();
        dataSets.add(setComp1);
        dataSets.add(setComp2);

        LineData data = new LineData(dataSets);


        XAxis xAxis = lineChart.getXAxis();
        xAxis.setGranularity(1f); // minimum axis-step (interval) is 1


        Description description = new Description();
        description.setText("Check daily expenses and incomes");
        lineChart.setDescription(description);

        lineChart.setData(data);
        lineChart.invalidate(); // refresh

    }

}

File: app/src/main/java/com/bernardo/maluleque/shibaba/ReportFragment.kt
package com.bernardo.maluleque.shibaba

import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import butterknife.BindView
import butterknife.ButterKnife
import com.bernardo.maluleque.shibaba.model.Transaction
import com.bernardo.maluleque.shibaba.utils.DateUtils
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.Description
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.*
import com.github.mikephil.charting.formatter.PercentFormatter
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import com.github.mikephil.charting.utils.ColorTemplate
import com.github.mikephil.charting.utils.MPPointF
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.*
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate
import java.util.stream.Collectors

class ReportFragment : Fragment() {
    @JvmField
    @BindView(R.id.pieChart)
    var pieChart: PieChart? = null

    @JvmField
    @BindView(R.id.lineChart)
    var lineChart: LineChart? = null

    @JvmField
    @BindView(R.id.barChart)
    var barChart: BarChart? = null
    private var mParam1 = 0
    private var dataMonth: String? = null
    private var db: FirebaseFirestore? = null
    private var registration: ListenerRegistration? = null
    private var user: FirebaseUser? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mParam1 = requireArguments().getInt(ARG_PARAM1)
            dataMonth = requireArguments().getString(ARG_PARAM2)
        }
        db = FirebaseFirestore.getInstance()
        user = FirebaseAuth.getInstance().currentUser
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_report, container, false)
        ButterKnife.bind(this, view)
        getTransactions(dataMonth)
        return view
    }

    fun getTransactions(dataMonth: String?) {
        val dateIntervals = DateUtils.getDateIntervals(dataMonth)
        val query = db!!.collection("users")
                .document(user!!.uid)
                .collection("transactions")
                .whereGreaterThanOrEqualTo("date", dateIntervals[0])
                .whereLessThanOrEqualTo("date", dateIntervals[1])
                .orderBy("date", Query.Direction.DESCENDING)
        registration = query.addSnapshotListener { queryDocumentSnapshots: QuerySnapshot?, e: FirebaseFirestoreException? ->
            if (e != null) {
                Log.e(TAG, "Listen failed.", e)
                return@addSnapshotListener
            }
            val transactions: MutableList<Transaction> = ArrayList()
            for (doc in queryDocumentSnapshots!!) {
                transactions.add(doc.toObject(Transaction::class.java))
            }
            if (transactions.isEmpty()) {
                return@addSnapshotListener
            }
            showData(transactions)
        }
    }

    private fun showData(transactions: List<Transaction>) {
        setPieChartData(transactions)
        setBarData(transactions)
        setLine2ChartData(transactions)
    }

    private fun setBarData(transactions: List<Transaction>) {
        val entries: MutableList<BarEntry> = ArrayList()
        val entriesExpense: MutableList<BarEntry> = ArrayList()
        val sumByExpenses = transactions.stream().filter { transaction: Transaction -> transaction.type == "Expense" }
                .collect(Collectors.groupingBy(Function { obj: Transaction -> obj.type }, Collectors.summingDouble { obj: Transaction -> obj.amount }))
        val sumByIncomes = transactions.stream().filter { transaction: Transaction -> transaction.type == "Income" }
                .collect(Collectors.groupingBy(Function { obj: Transaction -> obj.type }, Collectors.summingDouble { obj: Transaction -> obj.amount }))
        for ((_, value) in sumByExpenses) {
            entries.add(BarEntry(0f, value.toFloat()))
        }
        for ((_, value) in sumByIncomes) {
            entriesExpense.add(BarEntry(2f, value.toFloat()))
        }
        val set = BarDataSet(entries, "Total Income")
        set.color = Color.GREEN
        val set2 = BarDataSet(entriesExpense, "Total Expense")
        set2.color = Color.RED
        val data = BarData(set, set2)
        data.barWidth = 0.9f // set custom bar width
        val description = Description()
        description.text = "Totals expenses and incomes"
        barChart!!.description = description
        barChart!!.data = data
        barChart!!.setFitBars(true) // make the x-axis fit exactly all bars
        barChart!!.invalidate() // refresh
    }

    private fun setPieChartData(transactions: List<Transaction>) {
        val entries = ArrayList<PieEntry>()
        val sum = transactions.stream().filter { transaction: Transaction -> transaction.type == "Expense" }.collect(Collectors.groupingBy(Function { obj: Transaction -> obj.category }, Collectors.summingDouble { obj: Transaction -> obj.amount }))
        sum.entries.forEach(Consumer {
            entries.add(PieEntry(it.value.toFloat(), it.key))
        })
        val dataSet = PieDataSet(entries, "Categories")
        dataSet.setDrawIcons(false)
        dataSet.sliceSpace = 3f
        dataSet.iconsOffset = MPPointF(0f, 40f)
        dataSet.selectionShift = 5f

        // add a lot of colors
        val colors = ArrayList<Int>()
        for (c in ColorTemplate.VORDIPLOM_COLORS) colors.add(c)
        for (c in ColorTemplate.JOYFUL_COLORS) colors.add(c)
        for (c in ColorTemplate.COLORFUL_COLORS) colors.add(c)
        for (c in ColorTemplate.LIBERTY_COLORS) colors.add(c)
        for (c in ColorTemplate.PASTEL_COLORS) colors.add(c)
        colors.add(ColorTemplate.getHoloBlue())
        dataSet.colors = colors
        dataSet.selectionShift = 0f
        val data = PieData(dataSet)
        data.setValueFormatter(PercentFormatter(pieChart))
        data.setValueTextSize(11f)
        data.setValueTextColor(R.color.colorPrimaryText)
        pieChart!!.data = data

        // undo all highlights
        pieChart!!.highlightValues(null)
        pieChart!!.centerText = "Expenses by category"
        val description = Description()
        description.text = "Check where you expend the most"
        pieChart!!.description = description
        pieChart!!.invalidate()
    }

    private fun setLine2ChartData(transactions: List<Transaction>) {
        val sumByTypeAndDate = transactions.stream().filter { transaction: Transaction -> transaction.type == "Expense" }
                .collect(Collectors.groupingBy(Function { obj: Transaction -> obj.formattedDate }, Collectors.summingDouble { obj: Transaction -> obj.amount }))
        val sumByIncomeAndDate = transactions.stream().filter { transaction: Transaction -> transaction.type == "Income" }
                .collect(Collectors.groupingBy(Function { obj: Transaction -> obj.formattedDate }, Collectors.summingDouble { obj: Transaction -> obj.amount }))
        val values1 = ArrayList<Entry>()
        var ds = 0
        for ((_, value) in sumByTypeAndDate) {
            values1.add(Entry(ds.toFloat(), value.toFloat()))
            ds = ds + 1
        }
        val values2 = ArrayList<Entry>()
        for ((_, value) in sumByIncomeAndDate) {
            values2.add(Entry(ds.toFloat(), value.toFloat()))
            ds = ds + 1
        }
        val setComp1 = LineDataSet(values1, "Expenses")
        setComp1.axisDependency = YAxis.AxisDependency.LEFT
        setComp1.color = Color.RED
        setComp1.setCircleColor(Color.WHITE)
        val setComp2 = LineDataSet(values2, "Incomes")
        setComp2.axisDependency = YAxis.AxisDependency.LEFT
        setComp2.color = Color.GREEN
        setComp2.setCircleColor(Color.GREEN)
        val dataSets: MutableList<ILineDataSet> = ArrayList()
        dataSets.add(setComp1)
        dataSets.add(setComp2)
        val data = LineData(dataSets)
        val xAxis = lineChart!!.xAxis
        xAxis.granularity = 1f // minimum axis-step (interval) is 1
        val description = Description()
        description.text = "Check daily expenses and incomes"
        lineChart!!.description = description
        lineChart!!.data = data
        lineChart!!.invalidate() // refresh
    }

    companion object {
        private const val ARG_PARAM1 = "param1"
        private const val ARG_PARAM2 = "param2"
        private const val TAG = "ReportFragment"

        @JvmStatic
        fun newInstance(param1: Int, param2: String?): ReportFragment {
            val fragment = ReportFragment()
            val args = Bundle()
            args.putInt(ARG_PARAM1, param1)
            args.putString(ARG_PARAM2, param2)
            fragment.arguments = args
            return fragment
        }

        fun <T> distinctByKey(keyExtractor: Function<in T, *>): Predicate<T> {
            val seen: MutableSet<Any> = ConcurrentHashMap.newKeySet()
            return Predicate { t: T -> seen.add(keyExtractor.apply(t)) }
        }
    }
}